### DISCUSSION TITLE
`Module parse failed: Unexpected token` after upgrading an NPM dependency

### DISCUSSION THREAD

**Discussion by josemigallas** [babel/babel#16609] [Q&A]
Hi, our project is made with webpack 4 + babel 7 and we use `swagger-ui@3.52.5` from node_modules. 

We recently installed `swagger-ui@5.17.14` and something very weird happened when compiling:

```
ERROR in ./node_modules/swagger-ui/dist/swagger-ui-es-bundle-core.js 5507:11
Module parse failed: Unexpected token (5507:11)
File was processed with these loaders:
 * ./node_modules/babel-loader/lib/index.js
You may need an additional loader to handle the result of these loaders.
|     });
|   var operation_wrapper = (e, t) => class OperationWrapper extends Ke.default.Component {
>     onLoad = e => {
|       const {
|           operation: r
 @ ./app/javascript/packs/active_docs.ts 2:0-35 11:11-20
```

Usage:
```ts
import SwaggerUI from 'swagger-ui'
import 'swagger-ui/dist/swagger-ui.css'

SwaggerUI({ ... })
```

Webpack loader:
```
{
  test: /\.(js|jsx|mjs|ts|tsx)?(\.erb)?$/,
  include: [
    '/Users/jgallaso/projects/3scale/porta/app/javascript',
    '/Users/jgallaso/projects/3scale/porta/app/javascript/src'
  ],
  exclude: /node_modules/,
  use: [
    {
      loader: 'babel-loader',
      options: { cacheDirectory: true, cacheCompression: false, compact: false }
    }
  ]
}
```

Babel config:
```
module.exports = {
  plugins: [
  ],
  presets: [
    "@babel/preset-env",
    "@babel/preset-react"
  ],
}
```

Any ideas why is this failing? How to debug? What to try? We have to stick with webpack 4, which might be an issue...

**Comment by JLHwung** [comment]
You will have to apply babel-loader to the `./node_modules/swagger-ui/dist/swagger-ui-es-bundle-core.js` because webpack 4 does not support class properties , in your case it's `class { onLoad = ... }`.

Please replace `exclude: /node_modules/` by
```js
{
    exclude: {
      and: [/node_modules/], // Exclude libraries in node_modules ...
      not: [
        // Except for a few of them that needs to be transpiled because they use modern syntax
        /swagger-ui/,
      ]
    },
} 
```
See also https://github.com/babel/babel-loader/?tab=readme-ov-file#some-files-in-my-node_modules-are-not-transpiled-for-ie-11

If possible, please consider upgrade to webpack 5, as you are shipping transpiled code simply because webpack 4 does not recognize them, even if modern browsers support them natively.

**Reply by josemigallas** [reply]
Hi @JLHwung and thanks for the answer, it is super helpful. Unfortunately we can't upgrade to webpack 5😞.

I included `swagger-ui` to babel-loader and still **get the same error**. Then I added all the necessary plugins:
```js
plugins: [
  '@babel/plugin-transform-class-properties',
  '@babel/plugin-transform-optional-chaining',
  '@babel/plugin-transform-nullish-coalescing-operator',
  '@babel/plugin-transform-logical-assignment-operators'
]
```
Aren't these supposed to be included in preset-env? 🤔🤔🤔

Then the compilation broke with dozen of missing modules, such as this one:
```
ERROR in ./node_modules/swagger-ui/node_modules/swagger-client/es/execute/index.js
Module not found: Error: Can't resolve '@swagger-api/apidom-reference/configuration/empty' in '/Users/jgallaso/projects/3scale/porta/node_modules/swagger-ui/node_modules/swagger-client/es/execute'
 @ ./node_modules/swagger-ui/node_modules/swagger-client/es/execute/index.js 5:0-72 336:61-64
 @ ./node_modules/swagger-ui/dist/swagger-ui-es-bundle-core.js
 @ ./app/javascript/packs/active_docs.ts
 ```

Which I guess is now a webpack-related error.

**Reply by JLHwung** [reply]
> Aren't these supposed to be included in preset-env?

`preset-env` checks your `browserslist` config or the `targets` Babel option to determine required plugins. If `preset-env` does not enable class transforms, then you are already targeting to modern browsers (such as `latest 2 versions`) and they all supports ES2022 features such as class properties well. However, webpack 4 is 6 years old now, much older than any modern browsers, it unfortunately does not support class properties so you have to manually enable those plugins.

> Can't resolve '@swagger-api/apidom-reference/configuration/empty'

Does `swagger-client/es/execute` contain dynamic import `import()`? If so you may want to disable the `plugin-transform-dynamic-import` Babel plugin:

```js
["@babel/preset-env", { modules: false }]
```

**Reply by josemigallas** [reply]
Gotcha.

I don't think it has dynamic imports and `modules: false` didn't work. However, this reminded me of similar I had in the past, and problem was apparently package exports: https://github.com/swagger-api/swagger-js/issues/2921#issuecomment-1530982611.

Anyway... we decided to delay the swagger-ui upgrade until we are able to use webpack@5.

Thanks for you time and interest @JLHwung
